File: week4/java/Combinations_77.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        combine(n, k, 1, new ArrayList<Integer>(k), result);
        return result;
    }
    
    private void combine(int n, int k, int start, List<Integer> list, List<List<Integer>> result) {
        if (k == 0) {
            result.add(new ArrayList<Integer>(list));
            return;
        }
        
        for (int i = start; i <= n - k + 1; i++) {
            list.add(i);
            combine(n, k - 1, i + 1, list, result);
            list.remove(list.size() - 1);
        }
    }
}

File: week4/python/Combinations_77.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        
        # 1st approach
#         result = list()
        
#         def combination(combin: List[int], index: int):
#             if len(combin) == k:
#                 result.append(list(combin))
#                 return

#             if len(combin) > k:
#                 return
            
#             for i in range(index, n + 1):
#                 combin.append(i)
#                 combination(combin, i + 1)
#                 combin[:] = combin[:-1]
                
#         combination(list(), 1)
#         return result

        # 2nd approach
        import itertools
        return list(itertools.combinations(list(range(1,n+1)),k))
